File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        res = []
        queue = []
        if root is None:
            return res
        queue.append(root)
        zigzagFlag = True 
        while queue:
            size = len(queue)
            nodeLevel = [0] * size
            for i in range(size):
                node = queue.pop(0)
                index = i if zigzagFlag else (size - 1 - i)
                nodeLevel[index] = node.val
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
                
            zigzagFlag = not zigzagFlag
            res.append(nodeLevel)
            
        return res

File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        boolean left2Right = true;

        if (root == null) return result;
        while(queue.size() != 0) {
            List<Integer> level = new ArrayList<>();
            int queueLength = queue.size();

            for (int i = 0; i < queueLength; i++) {
                TreeNode node = queue.poll();

                level.add(node.val);

                if (node.left != null) queue.add(node.left);
                if (node.right != null) queue.add(node.right);
            }

            if (!left2Right) {
                Collections.reverse(level);
            }
            left2Right = !left2Right;
            result.add(level);
        }

        return result;
    }
}